rust/connlib/tunnel/src/device_channel/tun_windows.rs

```diff
@@ -336,14 +336,209 @@ fn set_iface_config(luid: wintun::NET_LUID_LH, mtu: u32) -> Result<()> {
     Ok(())
 }
 
+#[cfg(test)]
 mod tests {
-    /// Checks for regressions in issue #4765, un-initializing Wintun
+    use super::Tun;
+    use anyhow::Result;
+    use ip_packet::{
+        udp::MutableUdpPacket, IpPacket, MutableIpPacket, MutablePacket as _, Packet as _,
+        PacketSize as _,
+    };
+    use std::{
+        future::poll_fn,
+        net::{Ipv4Addr, Ipv6Addr},
+        time::Duration,
+    };
+    use tokio::{
+        net::UdpSocket,
+        time::{timeout, Instant},
+    };
+    use tracing_subscriber::EnvFilter;
+
+    /// Runs multiple Wintun tests.
+    ///
+    /// We don't have test names and UUIDs for the interfaces, and some are performance
+    /// tests, so we don't want to run these in parallel in multiple tests.
     #[test]
     #[ignore = "Needs admin privileges"]
-    fn resource_management() {
+    fn wintun() {
+        let _ = tracing_subscriber::fmt()
+            .with_env_filter(EnvFilter::from_default_env())
+            .with_test_writer()
+            .try_init();
+        perf().unwrap();
+        tunnel_drop();
+    }
+
+    /// Synthetic performance test
+    ///
+    /// Echoes UDP packets between a local socket and the Wintun interface
+    fn perf() -> Result<()> {
+        let rt = tokio::runtime::Runtime::new()?;
+        rt.block_on(async {
+            let mut tun = Tun::new()?;
+
+            const MTU: usize = 1_280;
+            const NUM_REQUESTS: u64 = 1_000;
+            const REQ_CODE: u8 = 42;
+            const REQ_LEN: usize = 1_000;
+            const RESP_CODE: u8 = 43;
+            const UDP_HEADER_SIZE: usize = 8;
+            const SERVER_PORT: u16 = 3000;
+
+            let ipv4 = Ipv4Addr::from([100, 90, 215, 97]);
+            let ipv6 = Ipv6Addr::from([0xfd00, 0x2021, 0x1111, 0x0, 0x0, 0x0, 0x0016, 0x588f]);
+            let mut device_manager =
+                connlib_shared::tun_device_manager::platform::TunDeviceManager::new()?;
+            device_manager.set_ips(ipv4, ipv6).await?;
+            tun.add_route(ipv4.into())?;
+
+            let server_addr = (ipv4, SERVER_PORT).into();
+
+            // Listen for incoming packets on Wintun, and echo them.
+            let server_task = tokio::spawn(async move {
+                tracing::debug!("Server task entered");
+                let mut requests_served = 0;
+                // We aren't interested in allocator speed or doing any processing,
+                // so just cache the response packet
+                let mut response_pkt = None;
+                let mut time_spent = Duration::from_millis(0);
+                loop {
+                    let mut req_buf = [0u8; MTU];
+                    poll_fn(|cx| tun.poll_read(&mut req_buf, cx)).await?;
+                    let start = Instant::now();
+                    // Copied from the DNS module in `firezone-tunnel`
```

Comment:
Please use the `ip_packet::make` module. The way we manipulate IP packets in the `dns` module is basically legacy code: https://github.com/firezone/firezone/issues/5540